File: app/controllers/Articles.java
package controllers;

import javax.persistence.Query;

import models.Article;
import play.data.Form;
import play.db.jpa.JPA;
import play.db.jpa.Transactional;
import play.mvc.BodyParser;
import play.mvc.Controller;
import play.mvc.Result;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectWriter;

public class Articles extends Controller {
	@Transactional(readOnly = true)
	@BodyParser.Of(BodyParser.Json.class)
	public static Result list() {
		ObjectMapper mapper = new ObjectMapper();
		String json = "";
		try {
			json = mapper.writeValueAsString(Article.list());
		} catch(Exception e) {
			// TODO
		}

		return ok(json);
	}

	@Transactional(readOnly = true)
	@BodyParser.Of(BodyParser.Json.class)
	public static Result show(String title) {
		Query query = JPA
				.em()
				.createQuery(
						"SELECT a FROM Article a WHERE a.title = :arg1 ORDER BY a.date DESC");
		query.setParameter("arg1", title);
		query.setMaxResults(1);
		Article article = (Article) query.getSingleResult();

		ObjectMapper mapper = new ObjectMapper();
		String json = "";
		try {
			json = mapper.writeValueAsString(article);
		} catch(Exception e) {
			// TODO
		}

		return ok(json);
	}

	static Form<Article> articleForm = Form.form(Article.class);

	@Transactional
	public static Result add() {
		Form<Article> filledForm = articleForm.bindFromRequest();
		if (filledForm.hasErrors()) {
			//return badRequest(create.render(filledForm));
			return TODO;
		} else {
			Article.create(filledForm.get());
			//return redirect(routes.Articles.index());
			return TODO;
		}
	}

	public static Result updateForm(String title) {
		return TODO;
	}

	@Transactional
	public static Result update(String title) {
		return TODO;
	}

	public static Result deleteForm(String title) {
		return TODO;
	}

	@Transactional
	public static Result delete(String title) {
		return TODO;
	}

	@Transactional
	public static Result delete(Long id) {
		Article.delete(id);
		//return redirect(routes.Articles.index());
		return TODO;
	}
	
	@Transactional(readOnly = true)
	public static Result history() {
		return TODO;
	}
}

File: app/controllers/Articles.scala
package controllers

import play.api.mvc._
import play.api.libs.json._
import play.api.db.slick._
import play.api.Play.current
import play.api.data._
import play.api.data.Forms._
import models.Article
import models.Revision
import java.util.Date

object Articles extends Controller with Secured {

	implicit object ArticleFormat extends Format[Article] {
		def reads(json: JsValue) = JsSuccess(Article(
			(json \ "id").as[Option[Long]],
			(json \ "title").as[String],
			(json \ "content").as[String],
			(json \ "date").as[Date]
		))

		def writes(a: Article): JsValue = {
			Json.obj(
				"id" -> a.id,
				"title" -> a.title,
				"content" -> a.content,
				"date" -> a.date
			)
		}
	}

	implicit object RevisionFormat extends Format[(Revision, String)] {
		def reads(json: JsValue) = JsSuccess((Revision(
			(json \ "id").as[Option[Long]],
			(json \ "title").as[String],
			(json \ "content").as[String],
			(json \ "articleId").as[Long],
			(json \ "userId").as[Long],
			(json \ "date").as[Date]), null
			))

		def writes(t: (Revision, String)): JsValue = {
			val r = t._1

			Json.obj(
				"id" -> r.id,
				"title" -> r.title,
				"content" -> r.content,
				"articleId" -> r.articleId,
				"userId" -> r.userId,
				"date" -> r.date,
				"userName" -> t._2
			)
		}
	}

	def list(keyword: Option[String]) = DBAction {
		implicit rs =>
			Ok(Json.toJson(
				models.Articles.list(keyword)))
	}

	def get(title: String) = DBAction {
		implicit rs =>
			Ok(Json.toJson(models.Articles.findByTitle(title)))
	}

	val articleForm = Form(
		tuple(
			"id" -> optional(longNumber),
			"title" -> nonEmptyText,
			"content" -> nonEmptyText
		)
	)

	def add = isAuthenticated {
		user => implicit rs =>
			val (_, title, content) = articleForm.bindFromRequest.get
			models.Articles.insert(new Article(title, content), user.id.get)
			Ok
	}

	def update(id: Long) = isAuthenticated {
		user => implicit rs =>
			val (_, title, content) = articleForm.bindFromRequest.get
			models.Articles.update(id, new Article(title, content), user.id.get)
			Ok
	}

	def delete(title: String) = DBAction {
		implicit rs =>
			models.Articles.delete(title)
			Ok
	}

	def revisions(title: String) = DBAction {
		implicit rs =>
			val articleId = models.Articles.findByTitle(title).get.id.get
			Ok(Json.toJson(
				models.Revisions.findByArticleId(articleId)))
	}
}
